test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -401,12 +402,70 @@ func CreateSnapshotResource(sDriver SnapshottableTestDriver, config *PerTestConf
 		// to pre-provision the snapshot as we normally would using their API.
 		// We instead dynamically take a snapshot and create another snapshot using
 		// the first snapshot's snapshot handle.
-		ginkgo.Skip("Preprovisioned test not implemented")
-		ginkgo.By("taking a snapshot with deletion policy retain")
-		ginkgo.By("recording the volume handle and status.snapshotHandle")
+		ginkgo.By("taking a dynamic snapshot")
+		r.Vs = getSnapshot(pvcName, pvcNamespace, r.Vsclass.GetName())
+		r.Vs, err = dc.Resource(SnapshotGVR).Namespace(r.Vs.GetNamespace()).Create(context.TODO(), r.Vs, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+
+		err = WaitForSnapshotReady(dc, r.Vs.GetNamespace(), r.Vs.GetName(), framework.Poll, framework.SnapshotCreateTimeout)
+		framework.ExpectNoError(err)
+
+		r.Vs, err = dc.Resource(SnapshotGVR).Namespace(r.Vs.GetNamespace()).Get(context.TODO(), r.Vs.GetName(), metav1.GetOptions{})
+
+		snapshotStatus := r.Vs.Object["status"].(map[string]interface{})
+		snapshotContentName := snapshotStatus["boundVolumeSnapshotContentName"].(string)
+		framework.Logf("received snapshotStatus %v", snapshotStatus)
+		framework.Logf("snapshotContentName %s", snapshotContentName)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("updating the snapshot content deletion policy to retain")
+		r.Vscontent, err = dc.Resource(SnapshotContentGVR).Get(context.TODO(), snapshotContentName, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+		r.Vscontent.Object["spec"].(map[string]interface{})["deletionPolicy"] = "Retain"
+
+		r.Vscontent, err = dc.Resource(SnapshotContentGVR).Update(context.TODO(), r.Vscontent, metav1.UpdateOptions{})
+		framework.ExpectNoError(err)
+
+		ginkgo.By("recording the volume handle and snapshotHandle")
+		snapshotHandle := r.Vscontent.Object["status"].(map[string]interface{})["snapshotHandle"].(string)
+		framework.Logf("Recording snapshot handle: %s", snapshotHandle)
+		csiDriverName := r.Vsclass.Object["driver"].(string)
+
 		ginkgo.By("deleting the snapshot and snapshot content") // TODO: test what happens when I have two snapshot content that refer to the same content
+		err = dc.Resource(SnapshotGVR).Namespace(r.Vs.GetNamespace()).Delete(context.TODO(), r.Vs.GetName(), metav1.DeleteOptions{})
+		framework.ExpectNoError(err)
+
+		ginkgo.By("checking the Snapshot has been deleted")
+		err = utils.WaitForNamespacedGVRDeletion(dc, SnapshotGVR, r.Vs.GetName(), r.Vs.GetNamespace(), framework.Poll, framework.SnapshotDeleteTimeout)
+		framework.ExpectNoError(err)
+
+		err = dc.Resource(SnapshotContentGVR).Delete(context.TODO(), r.Vscontent.GetName(), metav1.DeleteOptions{})
+		framework.ExpectNoError(err)
+
+		ginkgo.By("checking the Snapshot content has been deleted")
+		err = utils.WaitForGVRDeletion(dc, SnapshotContentGVR, r.Vscontent.GetName(), framework.Poll, framework.SnapshotDeleteTimeout)
```

Comment:
We update the vs content object delete policy to "retain" in line 424. Should delete of vs content fail here? i.e framework.ExpectError(err) ? Can you please elaborate?

Reply:
The deletion policy can be confusing, I was confused as well. If the deletion policy is retain on vscontent, when vs is deleted vscontent will not be deleted. Further, if the vscontent is manually deleted then the underlying snapshot resource will not be deleted.

Reply:
The policy will not prevent vscontent from being deleted.

Reply:
I see, can you note this behavior in a comment in the code

Reply:
Noted in the code at line 437